### PR TITLE
Refactoring for zero-shot standalone usecase

### PR DESCRIPTION
**Pull request by sungchul1**
### Summary

This PR includes:
- [x] Refactoring for zero-shot standalone usecase
- [x] Fix bug in cascade inference that used logits with 3 dim, not 4 dim
- [x] Update metric names to align with detection
- [x] Fix `segmentation fault` error by using `num_workers=0`
    - `Segmentation fault` error occurred during releasing threads on dataloader used at PTQ, but it seems to depend on datasets.
    - TODO : 135462

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/visual_prompting/zero_shot_segment_anything.py [hunk 1]
```diff
@@ -594,8 +596,8 @@ def _decide_cascade_results(
                 # all predicted masks were zero masks, ignore them.
                 return None, torch.zeros(masks.shape[-2:], device="cpu")
 
-            best_idx = torch.argmax(scores[0])
-        return logits[:, best_idx], masks[0, best_idx]
+            best_idx = int(torch.argmax(scores[0]))
```
Comment (vinnamkim) [inline src/otx/algo/visual_prompting/zero_shot_segment_anything.py]: Converting to Python integer is necessary? It is not desired because data transfer between devices may be required for this conversion.

File: src/otx/core/data/entity/visual_prompting.py [hunk 2]
```diff
@@ -193,10 +193,14 @@ def pin_memory(self) -> ZeroShotVisualPromptingBatchDataEntity:
         super().pin_memory()
         self.prompts = [
             [tv_tensors.wrap(prompt.pin_memory(), like=prompt) if prompt is not None else prompt for prompt in prompts]
```
Comment (vinnamkim) [inline src/otx/core/data/entity/visual_prompting.py]: Is this true? At L159, you declared `prompts: list[list[tv_tensors.TVTensor] | None]`. There seems no change that `prompt` can be `None`.

File: src/otx/core/data/entity/visual_prompting.py [hunk 3]
```diff
@@ -153,10 +153,10 @@ class ZeroShotVisualPromptingBatchDataEntity(OTXBatchDataEntity[ZeroShotVisualPr
...
     """
 
-    masks: list[tv_tensors.Mask]
-    labels: list[LongTensor]
-    polygons: list[list[Polygon]]
-    prompts: list[list[tv_tensors.TVTensor]]
+    masks: list[tv_tensors.Mask | None]
+    labels: list[LongTensor | None]
+    polygons: list[list[Polygon] | None]
+    prompts: list[list[tv_tensors.TVTensor] | None]
```
Comment (vinnamkim) [inline src/otx/core/data/entity/visual_prompting.py]: Why does everything be changed to be able to `None`?

**Comment by vinnamkim** [comment]
> * [x]  Fix `segmentation fault` error by using `num_workers=0`
>   
>   * `Segmentation fault` error occurred during releasing threads on dataloader used at PTQ, but it seems to depend on datasets.

If this is true, why the other tasks have not shown the same error? Did they already set `num_workers=0` while PTQ?

**Comment by sungchul1** [comment]
> > * [x]   Fix `segmentation fault` error by using `num_workers=0`
> >   
> >   * `Segmentation fault` error occurred during releasing threads on dataloader used at PTQ, but it seems to depend on datasets.
> 
> If this is true, why the other tasks have not shown the same error? Did they already set `num_workers=0` while PTQ?

No, other tasks use `num_workers=2`, but the error occurred during releasing threads on dataloader.
It is very difficult to check the exact cause, but for now I can update `num_workers` to resolve the current issue.

File: src/otx/core/data/entity/visual_prompting.py (hunk 3 above)
Comment (sungchul1) [reply src/otx/core/data/entity/visual_prompting.py]: For standalone `infer`, target images will be coming as inputs only.
This implementation is to consider the situation that other inputs are not coming.

File: src/otx/algo/visual_prompting/zero_shot_segment_anything.py (hunk 1 above)
Comment (sungchul1) [reply src/otx/algo/visual_prompting/zero_shot_segment_anything.py]: As you said, it is unnecessary operation, but it is intended to match type with `best_idx` which can be assigned at L584.
If you think it is unnecessary, I'll revert it.

File: src/otx/core/data/entity/visual_prompting.py (hunk 2 above)
Comment (sungchul1) [reply src/otx/core/data/entity/visual_prompting.py]: Edited at https://github.com/openvinotoolkit/training_extensions/pull/3090/commits/a261a5dd78908acec69972682afcb6a3211ac524.

File: src/otx/algo/visual_prompting/zero_shot_segment_anything.py (hunk 1 above)
Comment (vinnamkim) [reply src/otx/algo/visual_prompting/zero_shot_segment_anything.py]: Please revert it

File: src/otx/core/data/entity/visual_prompting.py (hunk 3 above)
Comment (vinnamkim) [reply src/otx/core/data/entity/visual_prompting.py]: If so, why not giving empty lists?

**Comment by vinnamkim** [comment]
> > > * [x]    Fix `segmentation fault` error by using `num_workers=0`
> > >   
> > >   * `Segmentation fault` error occurred during releasing threads on dataloader used at PTQ, but it seems to depend on datasets.
> > 
> > 
> > If this is true, why the other tasks have not shown the same error? Did they already set `num_workers=0` while PTQ?
> 
> No, other tasks use `num_workers=2`, but the error occurred during releasing threads on dataloader. It is very difficult to check the exact cause, but for now I can update `num_workers` to resolve the current issue.

Please create a ticket and assign it to someone who can resolve this problem to trace this problem. Then, add a todo comment to link the ticket in this PR.

File: src/otx/algo/visual_prompting/zero_shot_segment_anything.py (hunk 1 above)
Comment (sungchul1) [reply src/otx/algo/visual_prompting/zero_shot_segment_anything.py]: Done at https://github.com/openvinotoolkit/training_extensions/pull/3090/commits/918a22628539123fd12b905d83873268c3e42356.

File: src/otx/core/data/entity/visual_prompting.py (hunk 3 above)
Comment (sungchul1) [reply src/otx/core/data/entity/visual_prompting.py]: 🫨I didn't think about that, done at https://github.com/openvinotoolkit/training_extensions/pull/3090/commits/6d2a514b4560da8aa8e3349389e3188f1bba518f.

**Review by jaegukhyun — DISMISSED** [review]

**Comment by vinnamkim** [comment]
> Then, add a todo comment to link the ticket in this PR.

@sungchul2, Did you leave a todo comment?

**Comment by sungchul1** [comment]
> > Then, add a todo comment to link the ticket in this PR.
> 
> @sungchul2, Did you leave a todo comment?

I did.

**Comment by sungchul1** [comment]
TODO : 135462

**Comment by sungchul1** [comment]
> > Then, add a todo comment to link the ticket in this PR.
> 
> @sungchul2, Did you leave a todo comment?

Done at https://github.com/openvinotoolkit/training_extensions/pull/3090/commits/c9ede7840a93bea8df630423c71206d18a1ee884.

**Review by vinnamkim — APPROVED** [review]
